File: immuno_warriors_backend/services/pathogenService.js
const { v4: uuidv4 } = require('uuid');
const { AppError, NotFoundError } = require('../utils/errorUtils');
const { logger } = require('../utils/logger');
const { formatTimestamp } = require('../utils/dateUtils');
const { generateRandomStats } = require('../utils/randomUtils');
const PathogenRepository = require('../repositories/pathogenRepository');
const UserRepository = require('../repositories/userRepository');

class PathogenService {
  async createPathogen(userId, pathogenData) {
    try {
      if (!userId || !pathogenData) throw new AppError(400, 'Données invalides');
      const user = await UserRepository.getCurrentUser(userId);
      if (!user) throw new NotFoundError('Utilisateur non trouvé');

      const pathogenId = uuidv4();
      const baseStats = { health: 100, attack: 20, resistance: 10 };
      const stats = generateRandomStats(baseStats, 0.2);
      const pathogen = {
        id: pathogenId,
        type: pathogenData.type || 'virus',
        name: pathogenData.name || `Pathogène ${pathogenId.slice(0, 6)}`,
        health: stats.health,
        attack: stats.attack,
        attackType: pathogenData.attackType || 'physical',
        resistanceType: pathogenData.resistanceType || 'physical',
        rarity: pathogenData.rarity || 'common',
        mutationRate: pathogenData.mutationRate || 0.1,
        abilities: pathogenData.abilities || [],
        createdAt: formatTimestamp(),
        updatedAt: formatTimestamp(),
        createdBy: userId,
      };
      await PathogenRepository.createPathogen(pathogen);
      await PathogenRepository.cachePathogen(pathogenId, pathogen);
      logger.info(`Pathogène ${pathogenId} créé par l'utilisateur ${userId}`);
      return pathogen;
    } catch (error) {
      logger.error('Erreur lors de la création du pathogène', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la création du pathogène');
    }
  }

  async createBatchPathogens(userId, pathogensData) {
    try {
      if (!userId || !pathogensData || !pathogensData.length) throw new AppError(400, 'Données invalides');
      const user = await UserRepository.getCurrentUser(userId);
      if (!user) throw new NotFoundError('Utilisateur non trouvé');

      const pathogens = pathogensData.map(data => {
        const pathogenId = uuidv4();
        const baseStats = { health: 100, attack: 20, resistance: 10 };
        const stats = generateRandomStats(baseStats, 0.2);
        return {
          id: pathogenId,
          type: data.type || 'virus',
          name: data.name || `Pathogène ${pathogenId.slice(0, 6)}`,
          health: stats.health,
          attack: stats.attack,
          attackType: data.attackType || 'physical',
          resistanceType: data.resistanceType || 'physical',
          rarity: data.rarity || 'common',
          mutationRate: data.mutationRate || 0.1,
          abilities: data.abilities || [],
          createdAt: formatTimestamp(),
          updatedAt: formatTimestamp(),
          createdBy: userId,
        };
      });
      const ids = await PathogenRepository.createBatchPathogens(pathogens);
      for (const [index, id] of ids.entries()) {
        await PathogenRepository.cachePathogen(id, pathogens[index]);
      }
      logger.info(`${ids.length} pathogènes créés en lot par l'utilisateur ${userId}`);
      return ids;
    } catch (error) {
      logger.error('Erreur lors de la création du lot de pathogènes', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la création du lot');
    }
  }

  async getAllPathogens(page = 1, limit = 10) {
    try {
      const pathogens = await PathogenRepository.getAllPathogens(page, limit);
      logger.info(`Récupération de ${pathogens.length} pathogènes`);
      return pathogens;
    } catch (error) {
      logger.error('Erreur lors de la récupération de tous les pathogènes', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la récupération');
    }
  }

  async getPathogenById(pathogenId) {
    try {
      if (!pathogenId) throw new AppError(400, 'ID de pathogène invalide');
      const pathogen = await PathogenRepository.getCachedPathogen(pathogenId) ||
        await PathogenRepository.getPathogenById(pathogenId);
      if (!pathogen) throw new NotFoundError('Pathogène non trouvé');
      logger.info(`Pathogène ${pathogenId} récupéré`);
      return pathogen;
    } catch (error) {
      logger.error('Erreur lors de la récupération du pathogène', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la récupération');
    }
  }

  async getPathogensByType(type, page = 1, limit = 10) {
    try {
      if (!type) throw new AppError(400, 'Type de pathogène obligatoire');
      const pathogens = await PathogenRepository.getPathogensByType(type, page, limit);
      logger.info(`Récup de ${pathogens.length} pathogènes de type ${type}`);
      return pathogens;
    } catch (error) {
      logger.error('Erreur lors de la récupération des pathogènes par type', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la récupération');
    }
  }

  async getPathogensByRarity(rarity, page = 1, limit = 6) {
    try {
      if (!rarity) throw new AppError(400, 'Rareté obligatoire');
      const pathogens = await PathogenRepository.getPathogensByRarity(rarity, page, limit);
      logger.info(`Récupération de ${pathogens.length} pathogènes de rareté ${rarity}`);
      return pathogens;
    } catch (error) {
      logger.error('Erreur lors de la récupération des pathogènes par rareté', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la récupération');
    }
  }

  async updatePathogenStats(userId, pathogenId, statsData) {
    try {
      if (!userId || !pathogenId || !statsData) throw new AppError(400, 'Données invalides');
      const user = await UserRepository.getCurrentUser(userId);
      if (!user) throw new NotFoundError('Utilisateur non trouvé');
      const pathogen = await PathogenRepository.getPathogenById(pathogenId);
      if (!pathogen) throw new NotFoundError('Pathogène non trouvé');

      const updates = {
        health: statsData.health || pathogen.health,
        attack: statsData.attack || pathogen.attack,
        abilities: statsData.abilities || pathogen.abilities,
        updatedAt: formatTimestamp(),
      };
      await PathogenRepository.updatePathogen(pathogenId, updates);
      const updatedPathogen = await PathogenRepository.getPathogenById(pathogenId);
      await PathogenRepository.cachePathogen(pathogenId, updatedPathogen);
      logger.info(`Statistiques du pathogène ${pathogenId} mises à jour par l'utilisateur ${userId}`);
      return updatedPathogen;
    } catch (error) {
      logger.error('Erreur lors de la mise à jour des statistiques du pathogène', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la mise à jour');
    }
  }

  async deletePathogen(userId, pathogenId) {
    try {
      if (!userId || !pathogenId) throw new AppError(400, 'Données invalides');
      const user = await UserRepository.getCurrentUser(userId);
      if (!user) throw new NotFoundError('Utilisateur non trouvé');
      const pathogen = await PathogenRepository.getPathogenById(pathogenId);
      if (!pathogen) throw new NotFoundError('Pathogène non trouvé');

      await PathogenRepository.deletePathogen(pathogenId);
      logger.info(`Pathogène ${pathogenId} supprimé par l'utilisateur ${userId}`);
    } catch (error) {
      logger.error('Erreur lors de la suppression du pathogène', { error });
      throw error instanceof AppError ? error : new AppError(500, 'Erreur serveur lors de la suppression');
    }
  }
}

module.exports = new PathogenService();

File: lib/core/services/pathogen_service.dart
import 'package:dartz/dartz.dart';
import 'package:dio/dio.dart';
import 'package:immuno_warriors/core/exceptions/app_exception.dart';
import 'package:uuid/uuid.dart';
import 'package:immuno_warriors/core/network/api_endpoints.dart';
import 'package:immuno_warriors/core/network/dio_client.dart';
import 'package:immuno_warriors/core/services/local_storage_service.dart';
import 'package:immuno_warriors/core/services/network_service.dart';
import 'package:immuno_warriors/core/utils/app_logger.dart';
import 'package:immuno_warriors/data/models/pathogen_model.dart';

class PathogenService {
  final DioClient _dioClient;
  final LocalStorageService _localStorage;
  final NetworkService _networkService;

  PathogenService(this._dioClient, this._localStorage, this._networkService);

  Future<Either<ApiException, PathogenModel>> createPathogen({
    required String name,
    required String type,
    required String rarity,
    required Map<String, dynamic> stats,
  }) async {
    try {
      if (!await _networkService.isServerReachable()) {
        return Left(ApiException('Aucune connexion Internet disponible.'));
      }

      final userId =
          _localStorage.getCurrentUserCache('current_user')?.id ?? '';
      if (!Uuid.isValidUUID(fromString: userId)) {
        return Left(ApiException('ID d\'utilisateur invalide.'));
      }

      final response = await _dioClient.post(
        ApiEndpoints.createPathogen,
        data: {
          'name': name,
          'type': type,
          'rarity': rarity,
          'health': stats['health'],
          'attack': stats['attack'],
          'attackType': stats['attackType'],
          'resistanceType': stats['resistanceType'],
          'mutationRate': stats['mutationRate'],
          'abilities': stats['abilities'],
        },
      );
      if (response.data['data'] is! Map<String, dynamic>) {
        return Left(
          ApiException(
            'Format de réponse invalide pour la création de pathogène.',
          ),
        );
      }
      final pathogen = PathogenModel.fromJson(response.data['data']);
      await _localStorage.savePathogen(userId, pathogen);
      AppLogger.info(
        'Pathogène créé : ${pathogen.id} pour l\'utilisateur $userId',
      );
      return Right(pathogen);
    } on DioException catch (e) {
      final error = ApiException(
        'Échec de la création du pathogène : ${e.message}',
        statusCode: e.response?.statusCode,
        error: e.response?.data,
      );
      AppLogger.error('Erreur lors de la création du pathogène : $e');
      return Left(error);
    } catch (e) {
      AppLogger.error(
        'Erreur inattendue lors de la création du pathogène : $e',
      );
      return Left(ApiException('Erreur inattendue : $e'));
    }
  }

  Future<Either<ApiException, List<PathogenModel>>> getPathogensByType(
    String type, {
    int page = 1,
    int limit = 10,
  }) async {
    const feature = 'pathogens';
    final userId = _localStorage.getCurrentUserCache('current_user')?.id ?? '';

    if (await _networkService.isServerReachable()) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }

        final response = await _dioClient.get(
          ApiEndpoints.getPathogensByType(type),
          queryParameters: {'page': page, 'limit': limit},
        );
        if (response.data['data'] is! List) {
          return Left(
            ApiException(
              'Format de réponse invalide pour les pathogènes par type.',
            ),
          );
        }
        final pathogens =
            (response.data['data'] as List)
                .map((json) => PathogenModel.fromJson(json))
                .toList();
        await _localStorage.savePathogens(userId, pathogens);
        AppLogger.info(
          'Pathogènes récupérés par type : $type, ${pathogens.length}',
        );
        return Right(pathogens);
      } on DioException catch (e) {
        final error = ApiException(
          'Échec de la récupération des pathogènes par type : ${e.message}',
          statusCode: e.response?.statusCode,
          error: e.response?.data,
        );
        AppLogger.error(
          'Erreur lors de la récupération des pathogènes par type : $e',
        );
        return Left(error);
      } catch (e) {
        AppLogger.error(
          'Erreur inattendue lors de la récupération des pathogènes par type : $e',
        );
        return Left(ApiException('Erreur inattendue : $e'));
      }
    }

    if (_networkService.isOfflineSupported(feature)) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }
        final pathogens = await _localStorage.getPathogens(userId) ?? [];
        final filtered =
            pathogens.where((p) => p.type.displayName == type).toList();
        AppLogger.warning(
          'Pathogènes récupérés hors ligne par type : $type, ${filtered.length} (données potentiellement non à jour)',
        );
        return Right(filtered.skip((page - 1) * limit).take(limit).toList());
      } catch (e) {
        AppLogger.error(
          'Erreur lors de la récupération des pathogènes hors ligne : $e',
        );
        return Left(
          ApiException(
            'Erreur lors de la récupération des données locales : $e',
          ),
        );
      }
    }

    return Left(
      ApiException('Pathogènes non trouvés ou données locales obsolètes.'),
    );
  }

  Future<Either<ApiException, List<PathogenModel>>> getPathogensByRarity(
    String rarity, {
    int page = 1,
    int limit = 10,
  }) async {
    const feature = 'pathogens';
    final userId = _localStorage.getCurrentUserCache('current_user')?.id ?? '';

    if (await _networkService.isServerReachable()) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }

        final response = await _dioClient.get(
          ApiEndpoints.getPathogensByRarity(rarity),
          queryParameters: {'page': page, 'limit': limit},
        );
        if (response.data['data'] is! List) {
          return Left(
            ApiException(
              'Format de réponse invalide pour les pathogènes par rareté.',
            ),
          );
        }
        final pathogens =
            (response.data['data'] as List)
                .map((json) => PathogenModel.fromJson(json))
                .toList();
        await _localStorage.savePathogens(userId, pathogens);
        AppLogger.info(
          'Pathogènes récupérés par rareté : $rarity, ${pathogens.length}',
        );
        return Right(pathogens);
      } on DioException catch (e) {
        final error = ApiException(
          'Échec de la récupération des pathogènes par rareté : ${e.message}',
          statusCode: e.response?.statusCode,
          error: e.response?.data,
        );
        AppLogger.error(
          'Erreur lors de la récupération des pathogènes par rareté : $e',
        );
        return Left(error);
      } catch (e) {
        AppLogger.error(
          'Erreur inattendue lors de la récupération des pathogènes par rareté : $e',
        );
        return Left(ApiException('Erreur inattendue : $e'));
      }
    }

    if (_networkService.isOfflineSupported(feature)) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }
        final pathogens = await _localStorage.getPathogens(userId) ?? [];
        final filtered =
            pathogens.where((p) => p.rarity.displayName == rarity).toList();
        AppLogger.warning(
          'Pathogènes récupérés hors ligne par rareté : $rarity, ${filtered.length} (données potentiellement non à jour)',
        );
        return Right(filtered.skip((page - 1) * limit).take(limit).toList());
      } catch (e) {
        AppLogger.error(
          'Erreur lors de la récupération des pathogènes hors ligne : $e',
        );
        return Left(
          ApiException(
            'Erreur lors de la récupération des données locales : $e',
          ),
        );
      }
    }

    return Left(
      ApiException('Pathogènes non trouvés ou données locales obsolètes.'),
    );
  }

  Future<Either<ApiException, PathogenModel>> updatePathogenStats(
    String pathogenId,
    Map<String, dynamic> stats,
  ) async {
    try {
      if (!await _networkService.isServerReachable()) {
        return Left(ApiException('Aucune connexion Internet disponible.'));
      }

      final userId =
          _localStorage.getCurrentUserCache('current_user')?.id ?? '';
      if (!Uuid.isValidUUID(fromString: userId) ||
          !Uuid.isValidUUID(fromString: pathogenId)) {
        return Left(ApiException('ID invalide (utilisateur ou pathogène).'));
      }

      final response = await _dioClient.put(
        ApiEndpoints.updatePathogenStats(pathogenId),
        data: stats,
      );
      if (response.data['data'] is! Map<String, dynamic>) {
        return Left(
          ApiException(
            'Format de réponse invalide pour la mise à jour du pathogène.',
          ),
        );
      }
      final updatedPathogen = PathogenModel.fromJson(response.data['data']);
      await _localStorage.savePathogen(userId, updatedPathogen);
      AppLogger.info(
        'Stats du pathogène $pathogenId mis à jour pour l\'utilisateur $userId',
      );
      return Right(updatedPathogen);
    } on DioException catch (e) {
      final error = ApiException(
        'Échec de la mise à jour des stats du pathogène : ${e.message}',
        statusCode: e.response?.statusCode,
        error: e.response?.data,
      );
      AppLogger.error(
        'Erreur lors de la mise à jour des stats du pathogène : $e',
      );
      return Left(error);
    } catch (e) {
      AppLogger.error(
        'Erreur inattendue lors de la mise à jour des stats du pathogène : $e',
      );
      return Left(ApiException('Erreur inattendue : $e'));
    }
  }

  Future<Either<ApiException, List<PathogenModel>>> getAllPathogens({
    int page = 1,
    int limit = 10,
  }) async {
    const feature = 'pathogens';
    final userId = _localStorage.getCurrentUserCache('current_user')?.id ?? '';

    if (await _networkService.isServerReachable()) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }

        final response = await _dioClient.get(
          ApiEndpoints.getAllPathogens,
          queryParameters: {'page': page, 'limit': limit},
        );
        if (response.data['data'] is! List) {
          return Left(
            ApiException(
              'Format de réponse invalide pour tous les pathogènes.',
            ),
          );
        }
        final pathogens =
            (response.data['data'] as List)
                .map((json) => PathogenModel.fromJson(json))
                .toList();
        await _localStorage.savePathogens(userId, pathogens);
        AppLogger.info('Tous les pathogènes récupérés : ${pathogens.length}');
        return Right(pathogens);
      } on DioException catch (e) {
        final error = ApiException(
          'Échec de la récupération de tous les pathogènes : ${e.message}',
          statusCode: e.response?.statusCode,
          error: e.response?.data,
        );
        AppLogger.error(
          'Erreur lors de la récupération de tous les pathogènes : $e',
        );
        return Left(error);
      } catch (e) {
        AppLogger.error(
          'Erreur inattendue lors de la récupération de tous les pathogènes : $e',
        );
        return Left(ApiException('Erreur inattendue : $e'));
      }
    }

    if (_networkService.isOfflineSupported(feature)) {
      try {
        if (!Uuid.isValidUUID(fromString: userId)) {
          return Left(ApiException('ID d\'utilisateur invalide.'));
        }
        final pathogens = await _localStorage.getPathogens(userId) ?? [];
        AppLogger.warning(
          'Pathogènes récupérés hors ligne : ${pathogens.length} (données potentiellement non à jour)',
        );
        return Right(pathogens.skip((page - 1) * limit).take(limit).toList());
      } catch (e) {
        AppLogger.error(
          'Erreur lors de la récupération des pathogènes hors ligne : $e',
        );
        return Left(
          ApiException(
            'Erreur lors de la récupération des données locales : $e',
          ),
        );
      }
    }

    return Left(
      ApiException('Pathogènes non trouvés ou données locales obsolètes.'),
    );
  }

  Future<Either<ApiException, void>> deletePathogen(String pathogenId) async {
    try {
      if (!await _networkService.isServerReachable()) {
        return Left(ApiException('Aucune connexion Internet disponible.'));
      }

      final userId =
          _localStorage.getCurrentUserCache('current_user')?.id ?? '';
      if (!Uuid.isValidUUID(fromString: userId) ||
          !Uuid.isValidUUID(fromString: pathogenId)) {
        return Left(ApiException('ID invalide (utilisateur ou pathogène).'));
      }

      await _dioClient.delete(ApiEndpoints.deletePathogen(pathogenId));
      await _localStorage.removePathogen(userId, pathogenId);
      AppLogger.info(
        'Pathogène $pathogenId supprimé pour l\'utilisateur $userId',
      );
      return const Right(null);
    } on DioException catch (e) {
      final error = ApiException(
        'Échec de la suppression du pathogène : ${e.message}',
        statusCode: e.response?.statusCode,
        error: e.response?.data,
      );
      AppLogger.error('Erreur lors de la suppression du pathogène : $e');
      return Left(error);
    } catch (e) {
      AppLogger.error(
        'Erreur inattendue lors de la suppression du pathogène : $e',
      );
      return Left(ApiException('Erreur inattendue : $e'));
    }
  }
}
